Manually specialize parts of CRC32 implementation to speed them up debug mode

stackotter opened the pull request:
These changes are related to [my debug mode optimisation PR on for `swift-png`](https://github.com/tayloraswift/swift-png/pull/77) and have to be merged before that PR can be merged.

These optimisations corresponded to about a 20% improvement in debug mode PNG decoding at the time that I implemented them.

Please tag a new version (probably just a patch version?) once this is merged to that I can update my `swift-png` PR to point to it.

tayloraswift commented on Sources/CRC/CRC32.swift:
```diff
@@ -40,16 +40,90 @@ struct CRC32:Hashable, Sendable
+            return Self.table[index] ^ state >> 8
         }
     }
+
```
out of curiosity, can we get away with a `[UInt8]` typed overload that calls into the generic `Sequence<UInt8>` overload?

this ought to be enough to trigger specialization if i remember correctly

stackotter replied:
This is specifically a debug mode optimisation, and I don't think the compiler really does any specialisation in debug mode. Optimising for debug mode and optimising for release mode seem to be completely different beasts. As soon as a generic function is called, even if it's in the standard library, there seems to be a bunch of overhead. Here of course the overhead isn't at the call-site of that function, but it accumulates on every step of reduce due to the generic iterator and stuff.

tayloraswift replied:
okay

tayloraswift commented on Sources/CRC/CRC32.swift:
```diff
@@ -40,16 +40,90 @@ struct CRC32:Hashable, Sendable
+    ///
+    /// This manually specialized implementation is much faster in debug mode than the
+    /// generic implementation, but exactly the same in release mode.
+    @inlinable public
```
should this be `consuming`, and should there also be a generic `Sequence<UInt8>` overload?

stackotter replied:
There's already a generic overload above (which this manually specified method is kinda a copy of). I don't think `consuming` would make a difference to debug mode performance, but I do agree that it'd make sense for both the existing method and this method to have.

tayloraswift replied:
this type is bitwise copyable, so `consuming` would not affect performance, it is more a matter of “usage hint”. it’s also a bit simpler to spell the body of the function, as you simply mutate `self` and return it

please do add a `consuming func updated(with message:borrowing some Sequence<UInt8>)`, somebody (probably myself) will inevitably reach for it

also, since these are performance hooks, i would *really* prefer if the `[UInt8]` overloads were prefixed with an underscore. overloading *sucks* in Swift, and it would be nice if documentation authors (both in this package and downstream) could still refer to `update(with:)` without having to guess an FNV-1 hash :)

tayloraswift commented on Sources/CRC/CRC32.swift:
```diff
@@ -40,16 +40,90 @@ struct CRC32:Hashable, Sendable
         {
             (state:UInt32, byte:UInt8) in
-            Self.table[Int.init(UInt8.init(truncatingIfNeeded: state) ^ byte)] ^ state >> 8
+            let indexByte:UInt8 = UInt8.init(truncatingIfNeeded: state) ^ byte
```
since the performance-sensitive caller is going through the `[UInt8]` overload, do we actually need this?

stackotter commented:
Apologies for the long delay, things got pretty busy and then I moved on to other not png-related parts of my projects. I've now switched `swift-image-formats` over to using `libpng` because it provided the debug mode (and release mode) performance I require for my use case.

I'm pretty sure I've addressed your concerns with my latest commit. 

If this PR goes against the philosophy of this project (e.g. if debug mode performance is a non-goal) I don't mind if you decide to close this PR instead of merging. I'm not too attached either way.

tayloraswift commented:
looks good, thanks! sorry swift-png didn’t work out for you.
